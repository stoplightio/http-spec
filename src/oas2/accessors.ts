import { DeepPartial, Dictionary } from '@stoplight/types';
import { compact, get, isEmpty, isString, map, merge } from 'lodash';
import { negate } from 'lodash/fp';
import { Operation, Security, Spec } from 'swagger-schema-official';

export type SecurityWithKey = Security & { key: string };

export function getSecurities(
  spec: Partial<Spec>,
  operationSecurity: Array<Dictionary<string[], string>> | undefined,
): SecurityWithKey[][] {
  const globalSecurities = getSecurity(spec.security, spec.securityDefinitions || {});
  const operationSecurities = getSecurity(operationSecurity, spec.securityDefinitions || {});

  const securities = !!operationSecurity ? operationSecurities : globalSecurities;

  return securities.filter(negate(isEmpty));
}

export function getProduces(spec: DeepPartial<Spec>, operation: DeepPartial<Operation>) {
  return getProducesOrConsumes('produces', spec, operation);
}

export function getConsumes(spec: DeepPartial<Spec>, operation: DeepPartial<Operation>) {
  return getProducesOrConsumes('consumes', spec, operation);
}

function getSecurity(
  security: Array<Dictionary<string[], string>> | undefined,
  definitions: Dictionary<Security, string>,
): SecurityWithKey[][] {
  if (security === undefined) {
    return [];
  }
  return map(security, sec => {
    return compact(
      map(Object.keys(sec), (key: string) => {
        const def = definitions[key];
        if (def) {
          const defCopy: SecurityWithKey = merge<{ key: string }, Security>({ key }, def);
          return defCopy;
        }
        return null;
      }),
    );
  });
}

function getProducesOrConsumes(
  which: 'produces' | 'consumes',
  spec: DeepPartial<Spec>,
  operation: DeepPartial<Operation>,
): string[] {
  const mimeTypes = get(operation, which, get(spec, which, []));
  if (!Array.isArray(mimeTypes)) {
    return [];
  }

  return compact(mimeTypes).filter(v => v && isString(v));
}
