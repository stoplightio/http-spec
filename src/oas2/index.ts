export * from './operation';
export * from './service';
export * from './transformers/params';
