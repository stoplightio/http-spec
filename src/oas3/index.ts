export * from './operation';
export * from './service';
