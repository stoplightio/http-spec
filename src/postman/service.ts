import { Collection, CollectionDefinition, Item, ItemGroup } from 'postman-collection';
import { HttpServiceTransformer } from '../types';
import {
  isSecuritySchemeOutcomeEqual,
  isStandardSecurityScheme,
  SecuritySchemeOutcome,
  transformSecurityScheme,
} from './transformers/securityScheme';
import { traverseItemsAndGroups } from './util';

export const transformPostmanCollectionService: HttpServiceTransformer<CollectionDefinition> = collectionDefinition => {
  const collection = new Collection(collectionDefinition);
  const resolvedCollection = new Collection(collection.toObjectResolved({ variables: collection.variables }, []));

  const securitySchemeOutcomes: SecuritySchemeOutcome[] = [];
  let securitySchemeIdx = 0;

  function addSecurityScheme(outcome: SecuritySchemeOutcome) {
    if (!securitySchemeOutcomes.find(o => isSecuritySchemeOutcomeEqual(o, outcome))) {
      securitySchemeOutcomes.push(outcome);
    }
  }

  traverseItemsAndGroups(
    (resolvedCollection as unknown) as ItemGroup<Item>,
    item => {
      const auth = item.getAuth();
      if (auth) {
        const transformed = transformSecurityScheme(auth, type => `${type}-${securitySchemeIdx++}`);
        if (transformed) addSecurityScheme(transformed);
      }
    },
    itemGroup => {
      if (itemGroup.auth) {
        const transformed = transformSecurityScheme(itemGroup.auth, type => `${type}-${securitySchemeIdx++}`);
        if (transformed) addSecurityScheme(transformed);
      }
    },
  );

  return {
    id: resolvedCollection.id,
    name: resolvedCollection.name,
    version: resolvedCollection.version
      ? resolvedCollection.version.string
        ? resolvedCollection.version.string
        : `${resolvedCollection.version.major}.${resolvedCollection.version.minor}.${resolvedCollection.version.patch}-${resolvedCollection.version.prerelease}`
      : '1.0.0',
    description: resolvedCollection.description?.toString(),
    securitySchemes: securitySchemeOutcomes.filter(isStandardSecurityScheme).map(outcome => outcome.securityScheme),
  };
};
