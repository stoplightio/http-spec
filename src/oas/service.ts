import { isPlainObject } from '@stoplight/json';
import { DeepPartial, IHttpService } from '@stoplight/types';
import pickBy = require('lodash.pickby');
import { OpenAPIObject } from 'openapi3-ts';
import { Spec } from 'swagger-schema-official';

import { isBoolean, isNonNullable, isString } from '../guards';
import { TranslateFunction } from '../types';
import { hasXLogo } from './guards';
import { translateTagDefinition } from './tags';
import { translateLogo } from './transformers/translateLogo';

export const transformOasService: TranslateFunction<DeepPartial<OpenAPIObject> | DeepPartial<Spec>, [], IHttpService> =
  function () {
    const document = this.document;
    const id = String(document['x-stoplight']?.id);
    this.ids.service = id;
    this.parentId = id;

    const httpService: IHttpService = {
      id,

      version: document.info?.version ?? '',
      name: document.info?.title ?? 'no-title',

      ...pickBy(
        {
          description: document.info?.description,
          termsOfService: document.info?.termsOfService,
        },
        isString,
      ),

      ...pickBy(
        {
          contact: document.info?.contact,
        },
        isPlainObject,
      ),

      ...pickBy(
        {
          internal: document['x-internal'],
        },
        isBoolean,
      ),
    };

    if (isPlainObject(document.info) && hasXLogo(document.info)) {
      httpService.logo = translateLogo(document.info);
    }

    const tags = Array.isArray(document.tags)
      ? document.tags.map(translateTagDefinition, this).filter(isNonNullable)
      : [];

    if (tags.length > 0) {
      httpService.tags = tags;
    }

    return httpService;
  };
