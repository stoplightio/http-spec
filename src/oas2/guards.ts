import { isPlainObject } from '@stoplight/json';
import type { Response, Security, Tag } from 'swagger-schema-official';

import { isValidOas2Param, Oas2ParamBase } from '../oas/guards';

export function isSecurityScheme(maybeSecurityScheme: unknown): maybeSecurityScheme is Security {
  return isPlainObject(maybeSecurityScheme) && typeof maybeSecurityScheme.type === 'string';
}

export const isTagObject = (maybeTagObject: unknown): maybeTagObject is Tag => {
  return isPlainObject(maybeTagObject) && typeof maybeTagObject.name === 'string';
};

export const isResponseObject = (maybeResponseObject: unknown): maybeResponseObject is Response =>
  isPlainObject(maybeResponseObject) &&
  ('description' in maybeResponseObject ||
    'schema' in maybeResponseObject ||
    'headers' in maybeResponseObject ||
    'examples' in maybeResponseObject);

export function isValidScheme(scheme: unknown): scheme is 'http' | 'https' | 'ws' | 'wss' {
  return typeof scheme === 'string' && ['http', 'https', 'ws', 'wss'].includes(scheme);
}

export function isQueryParam(param: unknown): param is Oas2ParamBase & { in: 'query' } {
  return isValidOas2Param(param) && param.in === 'query';
}

export function isPathParam(param: unknown): param is Oas2ParamBase & { in: 'path' } {
  return isValidOas2Param(param) && param.in === 'path';
}

export function isHeaderParam(param: unknown): param is Oas2ParamBase & { in: 'header' } {
  return isValidOas2Param(param) && param.in === 'header';
}

export function isBodyParam(param: unknown): param is Oas2ParamBase & { in: 'body' } {
  return isValidOas2Param(param) && param.in === 'body';
}

export function isFormDataParam(param: unknown): param is Oas2ParamBase & { in: 'formData' } {
  return isValidOas2Param(param) && param.in === 'formData';
}
