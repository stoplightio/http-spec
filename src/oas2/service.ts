import { isPlainObject } from '@stoplight/json';
import { HttpSecurityScheme, IHttpOperation, IHttpOperationResponse, Optional, Reference } from '@stoplight/types';
import pickBy = require('lodash.pickby');

import { withContext } from '../context';
import { isNonNullable, isString } from '../guards';
import { createContext } from '../oas/context';
import { bundleResolveRef } from '../oas/resolver';
import { transformOasService } from '../oas/service';
import { translateToComponents } from '../oas/transformers/components';
import { translateSchemaObjectFromPair } from '../oas/transformers/schema';
import { Oas2HttpServiceBundle, Oas2HttpServiceTransformer, OasVersion } from '../oas/types';
import { ArrayCallbackParameters } from '../types';
import { entries } from '../utils';
import { normalizeProducesOrConsumes } from './accessors';
import { transformOas2Operations } from './operation';
import { translateToSharedParameters } from './transformers/params';
import { translateToResponse } from './transformers/responses';
import { translateToSingleSecurity } from './transformers/securities';
import { translateToServer } from './transformers/servers';
import { Oas2TranslateFunction } from './types';

export const bundleOas2Service: Oas2HttpServiceBundle = ({ document: _document }) => {
  const ctx = createContext(_document, bundleResolveRef);
  const { document } = ctx;

  const { securitySchemes, ...service } = transformOas2Service({ document, ctx });
  const produces = normalizeProducesOrConsumes(ctx.document.produces);
  const components = {
    ...translateToComponents.call(ctx, OasVersion.OAS2, {
      responses([key, value]): Optional<
        IHttpOperationResponse<true> | (Pick<IHttpOperationResponse<true>, 'code'> & Reference)
      > {
        return translateToResponse.call(ctx, produces, key, value);
      },
      definitions: translateSchemaObjectFromPair,
      securitySchemes: translateSecurityScheme,
    }),
    ...translateToSharedParameters.call(ctx, document),
  } as any;

  const operations = transformOas2Operations(document, ctx) as unknown as IHttpOperation<true>[];

  return {
    ...service,
    operations,
    components,
  };
};

export const transformOas2Service: Oas2HttpServiceTransformer = ({ document, ctx = createContext(document) }) => {
  const httpService = transformOasService.call(ctx);

  if (document.info?.license) {
    httpService.license = {
      ...document.info.license,
      name: document.info.license.name || '',
    };
  }

  const schemes = Array.isArray(document.schemes) ? document.schemes.filter(isString) : [];

  const servers = schemes.map(translateToServer, ctx).filter(isNonNullable);

  if (servers.length) {
    httpService.servers = servers;
  }

  const securitySchemes = entries(document.securityDefinitions).map(translateSecurityScheme, ctx).filter(isNonNullable);

  if (securitySchemes.length) {
    httpService.securitySchemes = securitySchemes;
  }

  const security = Array.isArray(document.security)
    ? document.security
        .flatMap(sec => {
          if (!isPlainObject(sec)) return null;

          return Object.keys(sec).map(key => {
            const ss = securitySchemes.find(securityScheme => securityScheme.key === key);
            if (ss && ss.type === 'oauth2') {
              const flows = {};
              for (const flowKey in ss.flows) {
                const flow = ss.flows[flowKey];
                flows[flowKey] = {
                  ...flow,
                  scopes: pickBy(flow.scopes, (_val: string, scopeKey: string) => {
                    const secKey = sec[key];
                    if (secKey) return secKey.includes(scopeKey);
                    return undefined;
                  }),
                };
              }

              return {
                ...ss,
                flows,
              };
            }

            return ss;
          });
        })
        .filter(isNonNullable)
    : [];

  if (security.length) {
    httpService.security = security;
  }

  return httpService;
};

const translateSecurityScheme = withContext<
  Oas2TranslateFunction<ArrayCallbackParameters<[name: string, scheme: unknown]>, Optional<HttpSecurityScheme>>
>(function ([key, definition]) {
  if (!isPlainObject(definition)) return;

  return translateToSingleSecurity.call(this, { ...definition, key });
});
