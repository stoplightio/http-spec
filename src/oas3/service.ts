import { HttpSecurityScheme, IHttpService, INodeTag, IServer } from '@stoplight/types';
import { compact, filter, flatMap, keys, map } from 'lodash';

import { Oas3HttpServiceTransformer } from '../oas/types';
import { isSecurityScheme, isTagObject } from './guards';
import { transformToSingleSecurity } from './transformers/securities';

export const transformOas3Service: Oas3HttpServiceTransformer = ({ document }) => {
  const httpService: IHttpService = {
    id: '?http-service-id?',
    version: document.info?.version ?? '',
    name: document.info?.title ?? 'no-title',
  };

  if (document.info?.description) {
    httpService.description = document.info.description;
  }

  if (document.info?.contact) {
    httpService.contact = document.info.contact;
  }

  if (document.info?.license) {
    httpService.license = {
      ...document.info.license,
      name: document.info.license.name || '',
    };
  }

  if (document.info?.termsOfService) {
    httpService.termsOfService = document.info.termsOfService;
  }

  const servers = compact<IServer>(
    map(document.servers, server => {
      if (!server) return null;

      return {
        name: document.info?.title ?? '',
        description: server.description,
        url: server.url ?? '',
      };
    }),
  );
  if (servers.length) {
    httpService.servers = servers;
  }

  const securitySchemes = compact<HttpSecurityScheme>(
    keys(document.components?.securitySchemes).map(key => {
      const definition = document?.components?.securitySchemes?.[key];
      return isSecurityScheme(definition) && transformToSingleSecurity(definition, key);
    }),
  );
  if (securitySchemes.length) {
    httpService.securitySchemes = securitySchemes;
  }

  const security = compact(
    flatMap(document.security, sec => {
      if (!sec) return null;

      return keys(sec).map(key => {
        return securitySchemes.find(securityScheme => {
          return securityScheme.key === key;
        });
      });
    }),
  ) as HttpSecurityScheme[];
  if (security.length) {
    httpService.security = security;
  }

  const tags = filter(document.tags, tag => isTagObject(tag)) as INodeTag[];
  if (tags.length) {
    httpService.tags = tags;
  }

  return httpService;
};
