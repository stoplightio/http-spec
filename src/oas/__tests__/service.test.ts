import { transformOas2Service } from '../../oas2/service';
import { transformOas3Service } from '../../oas3/service';

import * as oas2KitchenSinkJson from './fixtures//oas2-kitchen-sink.json';
import * as oas3KitchenSinkJson from './fixtures//oas3-kitchen-sink.json';

describe('oas service', () => {
  test('openapi v2', () => {
    expect(transformOas2Service({ document: oas2KitchenSinkJson as any })).toMatchSnapshot();
  });

  test('openapi v3', () => {
    expect(transformOas3Service({ document: oas3KitchenSinkJson as any })).toMatchSnapshot();
  });

  test('openapi v2 and v3 should transform in the same way', () => {
    const oas2MappedService = transformOas2Service({ document: oas2KitchenSinkJson as any });
    const oas3MappedService = transformOas3Service({ document: oas3KitchenSinkJson as any });
    expect(oas2MappedService).toEqual(oas3MappedService);
  });
});
