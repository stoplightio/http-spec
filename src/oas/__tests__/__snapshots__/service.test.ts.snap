// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`oas service openapi v2 1`] = `
Object {
  "contact": Object {
    "email": "apiteam@swagger.io",
  },
  "description": "This is a sample server Petstore server.  You can find out more about Swagger at [http://swagger.io](http://swagger.io) or on [irc.freenode.net, #swagger](http://swagger.io/irc/).  For this sample, you can use the api key \`special-key\` to test the authorization filters.",
  "id": "?http-service-id?",
  "license": Object {
    "name": "Apache 2.0",
    "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
  },
  "name": "Swagger Petstore",
  "securitySchemes": Array [
    Object {
      "description": undefined,
      "flows": Object {
        "implicit": Object {
          "authorizationUrl": "https://petstore.swagger.io/oauth/dialog",
          "scopes": Object {
            "read:pets": "read your pets",
            "write:pets": "modify pets in your account",
          },
        },
      },
      "key": "petstore_auth",
      "type": "oauth2",
    },
    Object {
      "description": undefined,
      "in": "header",
      "key": "api_key",
      "name": "api_key",
      "type": "apiKey",
    },
  ],
  "servers": Array [
    Object {
      "description": undefined,
      "name": "Swagger Petstore",
      "url": "https://petstore.swagger.io/v2",
    },
    Object {
      "description": undefined,
      "name": "Swagger Petstore",
      "url": "http://petstore.swagger.io/v2",
    },
  ],
  "tags": Array [
    Object {
      "description": "Everything about your Pets",
      "externalDocs": Object {
        "description": "Find out more",
        "url": "http://swagger.io",
      },
      "name": "pet",
    },
    Object {
      "description": "Access to Petstore orders",
      "name": "store",
    },
    Object {
      "description": "Operations about user",
      "externalDocs": Object {
        "description": "Find out more about our store",
        "url": "http://swagger.io",
      },
      "name": "user",
    },
  ],
  "termsOfService": "http://swagger.io/terms/",
  "version": "1.0.0",
}
`;

exports[`oas service openapi v3 1`] = `
Object {
  "contact": Object {
    "email": "apiteam@swagger.io",
  },
  "description": "This is a sample server Petstore server.  You can find out more about Swagger at [http://swagger.io](http://swagger.io) or on [irc.freenode.net, #swagger](http://swagger.io/irc/).  For this sample, you can use the api key \`special-key\` to test the authorization filters.",
  "id": "?http-service-id?",
  "license": Object {
    "name": "Apache 2.0",
    "url": "http://www.apache.org/licenses/LICENSE-2.0.html",
  },
  "name": "Swagger Petstore",
  "securitySchemes": Array [
    Object {
      "description": undefined,
      "flows": Object {
        "implicit": Object {
          "authorizationUrl": "https://petstore.swagger.io/oauth/dialog",
          "scopes": Object {
            "read:pets": "read your pets",
            "write:pets": "modify pets in your account",
          },
        },
      },
      "key": "petstore_auth",
      "type": "oauth2",
    },
    Object {
      "description": undefined,
      "in": "header",
      "key": "api_key",
      "name": "api_key",
      "type": "apiKey",
    },
  ],
  "servers": Array [
    Object {
      "description": undefined,
      "name": "Swagger Petstore",
      "url": "https://petstore.swagger.io/v2",
    },
    Object {
      "description": undefined,
      "name": "Swagger Petstore",
      "url": "http://petstore.swagger.io/v2",
    },
  ],
  "tags": Array [
    Object {
      "description": "Everything about your Pets",
      "externalDocs": Object {
        "description": "Find out more",
        "url": "http://swagger.io",
      },
      "name": "pet",
    },
    Object {
      "description": "Access to Petstore orders",
      "name": "store",
    },
    Object {
      "description": "Operations about user",
      "externalDocs": Object {
        "description": "Find out more about our store",
        "url": "http://swagger.io",
      },
      "name": "user",
    },
  ],
  "termsOfService": "http://swagger.io/terms/",
  "version": "1.0.0",
}
`;
